core/search/migrator.py

```diff
@@ -0,0 +1,171 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Iterable, List, Optional
+
+from core.search.revision import SearchSchemaRevision
+from core.search.revision_directory import SearchRevisionDirectory
+from core.search.service import SearchService, SearchServiceFailedDocument
+
+
+class SearchMigrationException(Exception):
+    """The type of exceptions raised by the search migrator."""
+
+    def __init__(self, fatal: bool, message: str):
+        super().__init__(message)
+        self.fatal = fatal
+
+
+class SearchDocumentReceiverType(ABC):
+    """A receiver of search documents."""
+
+    @abstractmethod
+    def add_documents(
+        self, documents: Iterable[dict]
+    ) -> List[SearchServiceFailedDocument]:
+        """Submit documents to be indexed."""
+
+    @abstractmethod
+    def finish(self) -> None:
+        """Make sure all changes are committed."""
+
+
+class SearchDocumentReceiver(SearchDocumentReceiverType):
```

Comment:
Is this an extra step, these behaviours seem to be all part of the SearchService interface

Reply:
The `SearchService` is supposed to be an absolute bare minimum stronger-typed interface to the Opensearch client. I've tried to separate mechanism and policy. The `SearchDocumentReceiver` and `SearchMigrationInProgress` classes are wrappers to give the code that needs to use them a much smaller interface to deal with ("start", "add documents", "finish").